File: src/services/battlesRegistry.ts
import type { Army, Battle } from '../types/index'
import {ArmiesRegistry} from '../services/armiesRegistry'
//Registry pattern
// Con este registro podemos vincular cada instancia de batalla a los historiales de sus participantes y así poder recuperarla para su 
// consulta, sin generar acoplamiento de los detalles de cada batalla con la clase Army. Además, se centraliza la gestión de la 
// información relativa a las batallas, y se protege su acceso sólo a los Armys cuyos historiales incluyan el battleId en esta línea:
// if(army?.historyBattle.includes(battleId))
export class BattleRegistry {

    private static battles = new Map<string, string>()

    static registryBattle(battle:Battle, contendersIds:string[]){
        const contenders = contendersIds.map((contenderId:string) => ArmiesRegistry.getArmy(contenderId)) as Army[]
        battle.id=Math.random().toString()
        contenders.map((contender:Army)=> {
            contender?.addBattle(battle.id)
            this.battles.set(battle.id, contender.getId())
        })
        
    }
    static getBattle(battleId:string, armyId:string){
        const army = ArmiesRegistry.getArmy(armyId)
        if(army?.hadBattle(battleId)){
            return this.battles.get(battleId)
        }
    }

}

File: army_challenge/services/battles_registry.py
from ..types import Army, Battle 
from ..services import ArmiesRegistry
from typing import Dict, Optional, List
from random import random
#Registry pattern
# Con este registro podemos vincular cada instancia de batalla a los historiales de sus participantes y así poder recuperarla para su 
# consulta, sin generar acoplamiento de los detalles de cada batalla con la clase Army. Además, se centraliza la gestión de la 
# información relativa a las batallas, y se protege su acceso sólo a los Armys cuyos historiales incluyan el battleId en esta línea:
# if(battleId in army.history_battle)
class BattleRegistry:

    __battles: Dict[str,str] = {}

    @staticmethod
    def registry_battle(battle: Battle, contendersIds: List[str]):
        contenders : List[Army] = [ArmiesRegistry.get_army(contender_id) for contender_id in contendersIds]
        battle.id=str(random())
        for contender in contenders:
            contender.historyBattle.append(battle.id)
            self.battles[battle.id] = contender.getId()

    @staticmethod
    def get_battle(battle_id:str, army_id:str):
        army = ArmiesRegistry.getArmy(army_id)
        if(battle_id in army.historyBattle):
            return this.battles.get(battle_id)
